resources/assets/js/markdown-editor/plugins/twitter.js

```diff
@@ -0,0 +1,289 @@
+        <label for="value">Tweet URL or ID</label>
+        <input id="value" name="value" type="text" class="te-alt-text-input" />
+        <div class="te-button-section">
+            <button type="button" class="te-ok-button">${i18n.get(
+                "OK"
+            )}</button>
+            <button type="button" class="te-close-button">${i18n.get(
+                "Cancel"
+            )}</button>
+        </div>
+    `;
+
+    popupContent
+        .querySelector("button.te-ok-button")
+        .addEventListener("click", async () => {
+            const input = popupContent.querySelector('[name="value"]');
+            const twitterCode = extractTwitterCode(input.value);
+            // This function will cache the html so we can get it on the replacer function
+            getAndCacheTwitterEmbedCode(twitterCode)
+                .then(() => {
+                    input.setAttribute("disabled", "disabled");
+                    editor.exec("twitter", twitterCode);
+                })
+                .catch((error) => {
+                    alert("Something went wrong!");
+                    console.error(error);
+                })
+                .then(() => {
+                    input.removeAttribute("disabled");
+                    input.value = "";
+                    editor.eventManager.emit("closeAllPopup");
+                });
+        });
+
+    popupContent
+        .querySelector("button.te-close-button")
+        .addEventListener("click", () => {
+            editor.eventManager.emit("closeAllPopup");
+        });
+
+    return popupContent;
+};
+
+/**
+ * Initialize UI
+ * @param {object} editor - Editor instance
+ * @param {int} MenuIndex - index inside the menu
+ * @param {object} svgIcon - Svg template
+ * @ignore
+ */
+const initPopup = (editor, menuIndex, svgIcon) => {
+    const name = "twitter";
+    const title = "Embed Tweet";
+    const tooltip = "Embed Tweet";
+
+    const popupContent = createPopupContent(editor);
+
+    createPopup(editor, name, menuIndex, svgIcon, popupContent, title, tooltip);
+};
+
+const convertHtmlToMarkdown = (html) => {
+    const regex = new RegExp(
+        `<img\\s+[^>]*src="twitter:([^"]*)"[^>]*>`,
+        "gmsi"
+    );
+
+    // Used to ensure that the root element is a DIV so its produce valid HTML
+    // This avoid some strange behaviours on the preview
+    const wrapInDivRegex = new RegExp(
+        `<([A-Z][A-Z0-9]*)[^>]*data-nodeid="([^"]*)"[^>]*>(((?!<\\1)[\\s\\S])*<img\\s+[^>]*src="twitter:[^"]*"[^>]*>.*?(<\/\\1|))<\/\\1>`,
+        "gmsi"
+    );
+    const validHTML = html.replace(
+        wrapInDivRegex,
+        '<div data-nodeid="$2">$3</div>'
+    );
+
+    let matches;
+    let replacemenent = validHTML;
+
+    while ((matches = regex.exec(validHTML)) !== null) {
+        // This is necessary to avoid infinite loops with zero-width matches
+        if (matches.index === regex.lastIndex) {
+            regex.lastIndex++;
+        }
```

Comment:
I see this being reused a lot, can it be moved to a generic method that is used in all placed?